crates/bevy_pbr/src/render/parallax_mapping.wgsl

```diff
@@ -0,0 +1,106 @@
+#define_import_path bevy_pbr::parallax_mapping
+
+fn sample_depth_map(uv: vec2<f32>) -> f32 {
+    return textureSample(depth_map_texture, depth_map_sampler, uv).r;
+}
+
+// An implementation of parallax mapping, see https://en.wikipedia.org/wiki/Parallax_mapping
+// Code derived from: https://web.archive.org/web/20150419215321/http://sunandblackcat.com/tipFullView.php?l=eng&topicid=28
+fn parallaxed_uv(
+    depth_scale: f32,
+    max_layer_count: f32,
+    max_steps: u32,
+    // The original interpolated uv
+    uv: vec2<f32>,
+    // The vector from the camera to the fragment at the surface in tangent space
+    Vt: vec3<f32>,
+) -> vec2<f32> {
+    if max_layer_count < 1.0 {
+        return uv;
+    }
+    var uv = uv;
+
+    // Steep Parallax Mapping
+    // ======================
+    // Split the depth map into `layer_count` layers.
+    // When Vt hits the surface of the mesh (excluding depth displacement),
+    // if the depth is not below or on surface including depth displacement (textureSample), then
+    // look forward (+= delta_uv) on depth texture according to
+    // Vt and distance between hit surface and depth map surface,
+    // repeat until below the surface.
+    //
+    // Where `layer_count` is interpolated between `1.0` and
+    // `max_layer_count` according to the steepness of Vt.
+
+    let view_steepness = abs(Vt.z);
+    // We mix with minimum value 1.0 because otherwise,
+    // with 0.0, we get a division by zero in surfaces parallel to viewport,
+    // resulting in a singularity.
+    let layer_count = mix(max_layer_count, 1.0, view_steepness);
```

Comment:
Maybe it’s not clear to the DirectX shader compiler that Vt is a unit vector so abs(Vt.z) is in the [0,1] range.

Reply:
But also, this suggests that at glancing angles, where Vt.z should be close to 0, the number of layers will be close to 1.